Readme.md

```diff
@@ -225,38 +263,175 @@ function BankAccount() {
+  US: 'en-US'
+};
 ```
 
-## Use UPPERCASE for Constants
+### Use UPPERCASE for Constants
 
 Constants should be declared as regular variables or static class properties,
 using all uppercase letters.
 
-Node.js / V8 actually supports mozilla's [const][const] extension, but
-unfortunately that cannot be applied to class members, nor is it part of any
-ECMA standard.
+*NOTE*: Use of ES6 keyword `const` is okay in NodeJS / V8 engines but
+discouraged until ES6 becomes more widely accepted. And, variablenames should
+still be be UPPERCASE_LIKE_THIS.
+
+Annotation with `@const` is recommended for all constant variables.
 
 *Right:*
 
 ```js
+/** @const */
 var SECOND = 1 * 1000;
 
+var MINUTE = 60 * SECOND; // Okay, but discouraged.
+
 function File() {
 }
+
+/** @const */
 File.FULL_PERMISSIONS = 0777;
 ```
 
 *Wrong:*
 
 ```js
-const SECOND = 1 * 1000;
+var second = 1 * 1000;
 
 function File() {
 }
+
 File.fullPermissions = 0777;
 ```
 
-[const]: https://developer.mozilla.org/en/JavaScript/Reference/Statements/const
+### File and Directories
+
+File and directory names should be lower-lisp-case, such as `a-awesome-dir`,
+`a-super-cool-file.js`, or `some-random-class.js`.
+
+For test files, a suffix of `_test` must be inserted right before `.js` file
+extension; e.g. `a-super-cool-file_test.js`, or `some-random-class_test.js`.
+
+File name and directory names should contain no punctuation except for `-` or
+`_`. (Prefer `-` to `_`).
+
+In addition, unit test files should mirror the directory structure of the source
+file except under a root `test` directory; e.g. `section-a/class-b.js` vs
+`test/section-a/class-b_test.js`.
+
+*Right:*
+
+```
+dev/section-a/subsection-2/big-giant-class.js # subsection is an english word.
+dev/section-beta/multi-sections/constants.js
+dev/section/special-section/mobile-api-routes.js
+
+test/section/special-section/mobile-api-routes_test.js
+```
+
+*Wrong:*
+
+```
+dev/sectionA/subSection2/BigGiantClass.js
+dev/sectionbeta/multiSections/Constants.js
+dev/section/special-section/mobileApiRoutes.js
+
+test/mobile-api-routes_test.js # Must match directory structure
+```
+
+### Properties and methods visibility
+
+Private methods and properties should be named with a `_` prefix. E.g.
+`_someVar`, `SomeClass.prototype._somePrivateMethod`, or `obj._privateProp`.
+
+### Optional or variable function parameters
+
+Optional function arguments start with `opt_`.
+
+Functions that take a variable number of arguments should have the last argument
+named `var_args`, it may never be referred in code. Use `arguments` array
+instead.
+
+*Right:*
+
+```js
+function someFunc(requiredArg, opt_arg) {
+  // opt_arg is optional, in the last set of arguments.
+}
+
+function variableFunc(requiredArg, var_args) {
+  hack(requiredArg);
+  hackMore(arguments);
+}
+```
+
+
+*Wrong:*
+
+```js
+function someFunc(opt_arg, requiredArg) {
+  // requiredArg cannot follow an optional argument.
+}
+
+function variableFunc(requiredArg, var_args) {
```

Comment:
I'm on the fence on this one. Never seen it before, not sure I understand the benefit -- is it to make it known that a function has variable # arguments? Seems like we could achieve that with comment with less overhead?